### DISCUSSION TITLE
Suggestions on supporting custom E2 service models and new RAN functions?

### DISCUSSION THREAD

**Discussion by gth828r** [srsran/srsRAN_Project#272] [General Help] [answered]
Hello,

I was wondering if there was any documentation discussing how to add support for new E2 service models? If there is no existing guidance on this yet then that is OK, but I just wanted to check if there are any high-level pointers before I set off on my own. Particular topics I am wondering about:
- How to get serialization code from an ASN.1 specification that will work with srsRAN_Project (what are the manual and automated steps)
- How to hook a new RAN function into the E2 agent

Thanks,
Tim

**Comment by pgawlowicz** [comment]
Hi, 

You can have a look into the `e2sm_kpm_du_meas_provider` [class](https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_du_meas_provider_impl.cpp), that provides the metrics for the E2 agent located in a DU.

Specifically, adding a new metric would require (Step 1) registering a new metric in the `e2sm_kpm_du_meas_provider` constructor, for example,  `DRB.PacketSuccessRateUlgNBUu` metric is registered [here](https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_du_meas_provider_impl.cpp#L40).
Then (in Step 2) you need to define a function that delivers the metric measurement data (e.g., `get_drb_ul_success_rate` [function](https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_du_meas_provider_impl.cpp#L310)).

Note that the metric definitions (Step 1) are then checked against the metric definition from ORAN specifications -- [here](https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_metric_defs.h) we have structured all ORAN metrics from O-RAN.WG3.E2SM-KPM-R003-v3.00 (that are mainly based on  3GPP TS 28.552).

So if you need a custom metric, then you need to add a new metric definition also there.

Best, 
Piotr

**Reply by pgawlowicz** [reply, marked as answer]
Ah, you actually ask about adding a new E2SM model. Unfortunately, currently, we do not have such guideline. However, we already have E2SM_KPM and E2SM_RC, so only E2SM_CCC is missing. We highly recommend extending those service models as they are the only models defined in ORAN specs. 

The ASN.1 code is generated automatically from ASN.1 specification from the related ORAN documents. 
If you need your own custom RAN function (service model), you need to define it (together with ASN.1 messages) to be compliant with E2SM speficication (i.e., O-RAN.WG3.E2SM-R003-v03.01)

**Reply by gth828r** [reply]
Hello, thank you for your response! If I define an E2SM-compliant ASN.1 service model specification, are there any suggestions on how to generate the corresponding C++ or C code from the ASN.1? In the mean time, I will also look to see if it is possible to implement what I'd like to test through standard E2SM-RC and bypass that issue altogether.

**Reply by pgawlowicz** [reply]
We do not have any recommendations for ASN.1 generation. We have our own ASN.1 compiler, but it is not publicly available.
